Spyglass Prow implementation, config, and deployment

paulangton opened the pull request:
Implementation of Spyglass in Deck, separated out from #8830. This PR should merge after both #8920 and #8921, and will fail until they merge. Contains the details of endpoints, authentication, announcement, front end, config, and actual viewer implementations.

The viewer implementations are not the focus of this PR, they are there to show functionality, not production-hardened readiness. For this reason, the deployment of Spyglass links in the main Deck UI are commented out.

/area prow
/assign @fejta @BenTheElder @stevekuznetsov @krzyzacy

fejta commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +431,164 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+		src := strings.TrimPrefix(r.URL.Path, "/view/")
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return Prowjob view page: ", elapsed)
```
Do we really want all this perf data logged at info? Seems like this would be better as debug or else as something exported to velodrome

paulangton replied:
This was something mentioned by steve on another PR, there is a TODO to export these metrics somewhere, I created #8956 for it.

paulangton commented on prow/spyglass/viewers/metadata/metadata_viewer.go:
```diff
@@ -0,0 +1,113 @@
+
+// ViewHandler creates a view for prow job metadata
+// TODO: os image,
+func ViewHandler(artifacts []viewers.Artifact, raw string) string {
```
Absolutely relevant and important, but I'm going to leave the polishing and battle-readying of the viewers for a separate PR as I'd like to focus on the core library and functionality for this one (Spyglass links in Deck are commented out for this reason)

fejta replied:
This comment is meant to apply to all the files you are adding, not just this one. All your non-trivial classes really ought to have unit tests. This is an important practice to develop.

paulangton commented on prow/spyglass/viewers/metadata/metadata_viewer.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// Started is used to mirror the started.json artifact
+type Started struct {
```
I was able to find a matching struct in `testgrid/cmd/updater/main.go`, this will be addressed in followup PRs targeting viewers

stevekuznetsov replied:
Also need to update the `initupload` and `sidecar` code with these

fejta commented:
Please pass tests and let me know when you're ready for the next round of comments on this one!

paulangton commented:
This branch will fail until #8921 merges, but I'll be sure to let everyone
know when it does.

On Wed, Aug 8, 2018, 8:36 PM Erick Fejta <notifications@github.com> wrote:

> Please pass tests and let me know when you're ready for the next round of
> comments on this one!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/8922#issuecomment-411626303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHcDDbICI6RzJ7a3t_qLBrl7AcwB-woYks5uO65XgaJpZM4VrkL8>
> .
>

fejta commented:
Maybe that will be true today hopefully? :smiling_imp:

krzyzacy commented on prow/spyglass/viewers/metadata/metadata_viewer.go:
```diff
@@ -0,0 +1,113 @@
+		Elapsed: metadataViewData.Finished.Timestamp.Sub(metadataViewData.Started.Timestamp),
+	}
+	metadataViewData.Derived = d
+	t := template.Must(template.New(fmt.Sprintf("%sTemplate", name)).Parse(tmplt))
```
we probably should move the template validation to a presubmit, or when registering the view, instead of panic in the middle of here?

paulangton replied:
as per @cjwagner 's comment I loaded it at init time, consistent with the pattern used in `badge.go`

krzyzacy commented:
(I'll take over and rebase)

paulangton commented:
Reopening this for further development on #9021

paulangton commented:
This PR will have two commits to separate personal code written from code written at work

paulangton commented:
Anyone know what to do about the CLA here? I have signed with my current email and the authors are correct for each commit.

fejta commented:
You need to keep your @google.com email associated with your account or else you need to replace them with ones you authored.

paulangton commented:
Got a response from CNCF, they said we could either override the presubmit or change the author. I went ahead and changed the author since that was suggested by repo owners. CLA passes. Please let me know if we should change it back.

cjwagner commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,201 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+		matches := []string{}
+		r, err := regexp.Compile(re)
+		if err != nil {
+			logrus.WithError(err).Error("Regexp failed to compile.")
```
We should definitely load the regexps at config load time like @stevekuznetsov suggested.
If we configure an invalid regexp we want the config load to fail so that we continue to use the value from the old config while we emit errors. This prevents downtime while we fix the config.

cjwagner commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,201 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+			}
+		}
+		for _, vName := range viewerNames {
+			viewerCache[vName] = matches
```
Should we still add viewers to the cache if matches is empty? In other words, do we want to process views even if no matching artifacts are found?

paulangton replied:
Empty views are excluded at the Spyglass stage, but yes we can also do it here.

cjwagner commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,201 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "failed to read body", http.StatusBadRequest)
```
I think failing to read the request body is a 5xx not a 4xx?

paulangton replied:
I wasnt able to find that information, everywhere I looked said use generic 400 bad request, any source on that?

cjwagner replied:
Nope, it just seems like it should be a server error if we fail to read the request body. I guess the 4xx is assuming that the request body is malformed as a opposed to a network read error?

If 400 is typical lets keep that.

BenTheElder replied:
Technically it's probably a 413 (payload too large) but we don't check what the error is so 400 makes sense

cjwagner commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,201 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+		viewReq := &spyglass.ViewRequest{}
+		err = json.Unmarshal(body, viewReq)
+		if err != nil {
+			http.Error(w, "failed to marshal request body", http.StatusBadRequest)
```
typo: unmarshal

cjwagner commented on prow/config/config.go:
```diff
@@ -215,8 +215,23 @@ type Sinker struct {
+	// artifact.ReadAll(). To exclude outlier artifacts, set this limit to
+	// expected file size + variance. To include all artifacts with high
+	// probability, use 2*maximum observed artifact size.
+	SizeLimit int64 `json:"size_limit,omitempty"`
```
Please specify the units for the size.

cjwagner commented on prow/spyglass/viewers/buildlog/buildlog_template.go:
```diff
@@ -0,0 +1,41 @@
+
+package buildlog
+
+const tmplt = `
```
Replace constant templates like these with the compiled versions so that they are resolved at init time instead of when the view is actually used. Its safer and a bit more efficient. Example: https://github.com/kubernetes/test-infra/blob/5a33c36fa0440580f53d21b7aa3ba319e0a51aec/prow/cmd/deck/badge.go#L30-L51

paulangton replied:
Interesting, I like it. Added in latest.

paulangton commented:
I think I've addressed all reviews excluding specific viewers (will be addressed in follow-up PRs), what do we think about moving towards merging this?

krzyzacy commented on prow/ANNOUNCEMENTS.md:
```diff
@@ -1,6 +1,8 @@
-New features added to each components:
+New features added to each component:
+ - *July 26, 2018* `spyglass` is a pluggable viewing framework for artifacts
+   produced by prowjobs. See a demo [here](prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121)!
```
nit: +https to the link, or else it's a relative path

krzyzacy commented:
/lint

@cjwagner @fejta @stevekuznetsov any more comments?

paulangton commented on prow/spyglass/viewers/junit/junit_template.go:
```diff
@@ -0,0 +1,124 @@
+limitations under the License.
+*/
+
+package junit
```
This lint warning is super annoying, is there an open issue for this? I have a package comment elsewhere in the junit package.

krzyzacy replied:
feel free to open one :-)

paulangton replied:
#9118 :')

BenTheElder replied:
thanks! this lint warning is indeed very annoying

paulangton commented:
Addressed lint comments :+1:

stevekuznetsov commented on prow/ANNOUNCEMENTS.md:
```diff
@@ -1,6 +1,8 @@
-New features added to each components:
+New features added to each component:
+ - *July 26, 2018* `spyglass` is a pluggable viewing framework for artifacts
+   produced by prowjobs. See a demo [here](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121)!
```
this is still a 404

paulangton replied:
addressed in your last review

stevekuznetsov commented on prow/cluster/deck_deployment.yaml:
```diff
@@ -46,6 +46,7 @@ spec:
         - --redirect-http-to=prow.k8s.io
         - --oauth-url=/github-login
         - --job-config-path=/etc/job-config
+        - --spyglass=true
```
Are we ready for this?

paulangton replied:
There are no front end Deck UI components added yet, thus most users will still go straight to gubernator for viewing jobs. This is only to enable spyglass links for people who would like to use them. Deck UI components for fully adopting spyglass will be added once the viewers are production-ready.

krzyzacy replied:
I think we'll need to add the flag after the image bump, or else the deployment will be invalid. Or we can bump right after this is merged

stevekuznetsov commented on prow/cmd/deck/main.go:
```diff
@@ -172,6 +186,19 @@ func prodOnlyMain(configAgent *config.Agent, o options, mux *http.ServeMux) *htt
+		//TODO: Need to support authenticated buckets, #8910
+		c, err := storage.NewClient(context.Background(), option.WithoutAuthentication())
+		if err != nil {
+			logrus.WithError(err).Fatal("Error getting GCS client")
```
Why do we just continue on without a GCS client?

paulangton replied:
This is the way that the other two clients are handled in `prodOnlyMain`.

stevekuznetsov commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,202 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		r, err := regexp.Compile(re)
```
I would like to see us do this at config-load time

paulangton replied:
It is done at config-load time

stevekuznetsov commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,202 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+	viewsStart := time.Now()
+	lenses := sg.Views(viewerCache)
+	viewsElapsed := time.Since(viewsStart)
+	logrus.WithField("duration", viewsElapsed).Info("Loaded spyglass views")
```
we're going to want identifying fields on these -- what was the request, what were the parameters, etc

stevekuznetsov commented on prow/cmd/deck/main.go:
```diff
@@ -408,6 +435,202 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+		Source:      src,
+		ViewerCache: viewerCache,
+	}
+	t, err := template.ParseFiles("/static/template/spyglass-template.html")
```
this shouldn't be a hard-coded path

stevekuznetsov replied:
use `o.staticFilesLocation`

stevekuznetsov commented on prow/config/config.go:
```diff
@@ -215,8 +215,23 @@ type Sinker struct {
+	// read in entirety. This will only affect viewers attempting to use
+	// artifact.ReadAll(). To exclude outlier artifacts, set this limit to
+	// expected file size + variance. To include all artifacts with high
+	// probability, use 2*maximum observed artifact size.
```
Does this assume a normal distribution of artifact sizes? ;)

paulangton replied:
It does. If you'd like to collect some metrics on test artifact sizes that would be great! Until we have that data (and perhaps even with that data...) I'd prefer to mention only that `size_limit` is dependent on both your use case and the distribution of your test artifact sizes. Unfortunately that gives no real concrete advice to the administrator of the system, so I settled on this.

stevekuznetsov commented on prow/config/config.go:
```diff
@@ -737,6 +752,19 @@ func parseProwConfig(c *Config) error {
+	}
+
+	for k := range c.Deck.Spyglass.Viewers {
+		_, err := regexp.Compile(k)
```
we can store the compiled regex in a private field and not re-compile this every request, we do this for e.g. `trigger`

paulangton replied:
I looked through the code in `trigger`, I dont think that exact pattern can be adapted for this case because we have an arbitrary number of regexps. Instead I added a field to the spyglass config struct that caches the compiled regexps, which should have the same effect.

stevekuznetsov commented on prow/spyglass/viewers/junit/junit_template.go:
```diff
@@ -0,0 +1,124 @@
+	"html/template"
+)
+
+var junitTemplateText = `{{$numF := len .Failed}}
```
I don't think we have talked about this yet in earnest ... I really think iterating on this code will be a nightmare. It's fairly complex and the way we have it checked in, no editor will be using appropriate syntax highlighting for it, etc -- can we send HTML + CSS + JS and a data payload so at least the generation of the HTML is in JS and self-sufficient?

paulangton replied:
Agreed, the Javascript problem is brought up in #8967. In terms of iterating, I am very much open to better suggestions, but as I've said before, I wont be addressing viewer-specific comments on this PR. I think each of the viewers is important enough to warrant their own PRs (and this PR does not profess Spyglass as the canonical way to view jobs in Prow), so I'll leave this and other comments on viewers for later.

stevekuznetsov replied:
I don't think agreeing that a large problem exists but deciding to check in the code and punt on solving it is a reasonable approach. What if you decide you don't want to work on this anymore once this PR merges but before any future ones do? Let's not check in code that we know will require significant thought and potential re-work in thefuture.

paulangton replied:
Ok, let's see what we can do. The main problem that I'm encountering here is that If I put the HTML/CSS/JS in separate files their paths need to be relative to the `deck` deployment (likely the static files location), which ruins the self-contained nature of the viewers. I'm not aware of a pattern that allows me to get around this.

Additionally, reviewing code from the standpoint of "what if you dont want to work on this anymore" is not helpful.

cjwagner replied:
Is a lack of syntax highlighting the only problem you are describing here or is there more to it than that? If the problem is just that editors won't highlight syntax properly that doesn't seem like a large or blocking problem.
If there is some deeper problem that makes these viewers untenable then I agree that we should resolve it here or omit the viewers from the PR until they are suitable for merge.

BenTheElder replied:
FWIW many editors will syntax highlight mixed source code.. but go templates are only valid syntax as go templates. They are not guaranteed to be valid HTML prior to execution, nor should they. It is also quite useful to pre-render to HTML instead of sending data + static html + js for many use cases.

I'm happy to edit this code in the future, and I do think it is preferable to keep viewers self contained. I don't think this will be troublesome to rework, this is pretty standard. EG `hugo` which is popular (and used by our docs!) is basically a ton of stuff on top of go templates, and all docs files are essentially go template-enabled markdown.

cjwagner replied:
> Additionally, reviewing code from the standpoint of "what if you dont want to work on this anymore" is not helpful.

I think it kind of is. It is important to leave the master branch in a clean, working state. We don't want to commit half implemented or unmaintainable code so it is usually best to merge only the components that are complete and omit components that aren't complete until they are sufficiently polished (barring an emergency).
If the only problem here is syntax highlighting though, these viewers seem sufficiently complete to merge to me.

stevekuznetsov replied:
Every single time I have worked with a system where one language was embedded in another language or configuration format, it has been a nightmare. Developing new views and iterating on them would be frustrating and error-prone. It would be simple to check something in that looks reasonable on review but is subtly broken. All of my experience just points to this being an approach that is fraught with error and if we can design something that doesn't lead us down that path I think we will be much better off.

BenTheElder replied:
> It would be simple to check something in that looks reasonable on review but is subtly broken. 

This is true of all HTML. Validating HTML is non-trivial. Add in HTML + javascript rewriting it at runtime + some data format and it's definitely going to be subtly broken, as evidenced by every existing page in prow :-) Embedding a little data with a go template on the other hand is pretty easy understood compared to JS manipulating HTML, and not really any worse off on the HTML front.

New viewers do not need to use these templates, ideally we will quickly add external viewer tooling and viewer binaries can do whatever the maintainers feel are appropriate.

paulangton replied:
I'm open to doing the leg work in implementing whatever approach we feel is best, but to me it seems that a go template is the best way to go here, at least until we get external viewers. @stevekuznetsov thoughts?

stevekuznetsov replied:
If I'm the lone outsider on the issue, I'm ok with the Go template approach. If we find that it's frustrating to iterate on it, we can revisit in the future. I just wanted to make sure we considered the issue at hand explicitly before merging this PR.

krzyzacy replied:
I'd say let's not worry about this for now (until someone complains?), or feel free to open a tracking issue

stevekuznetsov commented on prow/spyglass/viewers/metadata/metadata_viewer.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// Finished is used to mirror the finished.json artifact
+type Finished struct {
```
You're going to want `omitempty` on everything but the most basic -- otherwise you're going to create #6092 for your viewer. Your downstream logic for handling these should handle empty values as well.

stevekuznetsov commented on prow/spyglass/viewers/metadata/metadata_viewer.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// Derived contains metadata derived from provided metadata for insertion into the template
+type Derived struct {
```
Love this

paulangton commented:
I believe I've addressed all outstanding feedback :+1:

paulangton commented:
👍 will remove hold after another deployment test and rebase

paulangton commented:
should be all set!

paulangton commented:
/lgtm 
/meow

paulangton commented:
/woof produces far superior images

krzyzacy commented:
/lgtm

paulangton commented:
/hold cancel
